### PR TITLE
feat(usage): Report resource duration. Closes #1066

### LINKED ISSUE #1066: Feature Request:   Calculate workflow resource usage
For every workflow, many users want to know how many resource(cpu, memory) is used.   This info is very useful for user to compute costs.  We can use pod resource requests or limit to show every pod resource usage.   

Maybe `argo get <workflow>` can show info like this.     Can someone accept this feature?  I can raise a PR to this.

Example:

```
➜  argo git:(release-2.2) argo get hello-world-gdpwz
Name:                hello-world-gdpwz
Namespace:           default
ServiceAccount:      default
Status:              Succeeded
Created:             Wed Oct 31 15:01:27 +0800 (4 minutes ago)
Started:             Wed Oct 31 15:01:27 +0800 (4 minutes ago)
Finished:            Wed Oct 31 15:01:50 +0800 (3 minutes ago)
Duration:            23 seconds
Total CPU:           0.00767    (core*hour)
Total Memory:        0.00639    (GB*hour)

STEP                  PODNAME            DURATION  MESSAGE  CPU(core*hour)  MEMORY(GB*hour)
 ✔ hello-world-gdpwz  hello-world-gdpwz  23s                0.00767         0.00639
```

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlike to be merged.
* [x] Optional. I've added My organization is added to the README.
* [x] I've signed the CLA and required builds are green. 

Closes #1066

Screen Shot 2020-02-12 at 4 47 01 PM

Screen Shot 2020-02-12 at 4 47 32 PM

As it stands, this computes the usage and stores in it the status field for pods only. It then provides a method to get the total usage.

The usages of steps/DAGs (which would be the sum of the pods within them) is not calculated, but could be useful.

### CODE REVIEW

**Comment by ddseapy** [comment]
So my understanding from reading this, is that a user configures a "utilization per resource" in the workflow: how much they want to count for each resource per minute  (ie. 1Gi RAM/min = 4 utilization points, 1 CPU/min = 6 utilization points).

Argo looks at duration between container start/end times, along with the container's resource request.  It takes that and the user's "utilization per resource" to compute an integer that is representative of the utilization.

If this is correct, then this satisfies our use case (other than needing gpu resource type, which I can create a subsequent merge request if that makes sense).

**Comment by ddseapy** [comment]
This looks great @alexec, and thanks for adding the gpu metric too!  I'm happy to test this out whenever you think it's ready.

File: docs/usage.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Usage
+
+![alpha](assets/alpha.svg)
+
+Argo Workflows estimates how much resource your workflow has used and save this information. This is know as "resource usage" and is intended as an **indicative but not accurate** value.
```
Comment (ddseapy) [inline docs/usage.md]: nitpick: `is known as`

File: docs/usage.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+* Memory usage: 3 * 60s * 100m / 1Gi = 0s*memory
+* GPU usage: 2 * 60s * = 3m*nvidia.com/gpu
+
+## Limitations & Assumptions
+
+To calculate the usage we assume that request/limit/default for a resource is a good enough representative of the pods average usage.
+
+This is **never** actually the case:
+
+* The pod will probably use more that that the request and less than the limit.
```
Comment (ddseapy) [inline docs/usage.md]: nitpick: rm `that that`

File: docs/usage.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+* The pod may use more than the limit or less than the request.
```
Comment (ddseapy) [inline docs/usage.md]: wouldn't kubernetes kill the pod if it used more than the limit?

Comment (alexec) [reply docs/usage.md]: not as I understand it, instead you go to the top of the list to be killed, you may or may not be killed

Comment (ddseapy) [reply docs/usage.md]: ah, i didn't know that.  That makes sense.

**Comment by alexec** [comment]
@jessesuen this is a PoC for creating a **indicative but not accurate** estimate of resource usage.

This is a white-space project.

I'm in two minds about it. While I like the indicative nature of it, and the cheap cost of implementation, I'm quite aware that the way the usage is estimate is only really very rough. 

The question therefore in my mind is this  - *is the benefit of this metric greater or less than the cost of people miss-understanding it*?

Thoughts?

**Comment by alexec** [comment]
> The question therefore in my mind is this - _is the benefit of this metric greater or less than the cost of people miss-understanding it_?

One mitigation to this is to change it from "usage" to "guesstimate usage".

**Comment by alexec** [comment]
Lets call a spade "a spade". 

`/spec/nodes/-/usageIndicator.`

**Comment by hnykda** [comment]
As a naive oberver and lurker, but a user of argo, a rough estimate is better than no estimate _if_ I know it's rough. So if it's well documented (and maybe with a tooltip), it's better to have it.

File: docs/usage-indicator.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+Usage indicator calculation is always an estimate, and is calculated by [../util/usage/estimator.go](../util/usage/estimator.go) based on container duration, specified pod resource requests, limits, or defaults (for memory and CPU). 
+
+Each indicator is divided by a common denominator depending or resource type.
+
+### Example
+
+A pod that runs for 3m, with a CPU limit of 2000m, no memory request and an `nvidia.com/gpu` resource limit of 1:
+
+* CPU: 3 * 60s * 2000m / 1000m = 6m*cpu
+* Memory: 3 * 60s * 100m / 1Gi = 0s*memory
```
Comment (simster7) [inline docs/usage-indicator.md]: This makes no sense to me.

What is `m` in this context? Milicores? Mebibytes?

What is a `memory` unit defined as? 

If I take this expression literally, it says to me either "18,000 second-milicores per Gigibyte" or "180 seconds per kibibye", depending on what `m` is here. Neither of these make sense per se or as a measure of memory usage.

How do we end up at "0 second memories" here?

File: docs/usage-indicator.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+* GPU: 2 * 60s * = 3m*nvidia.com/gpu
```
Comment (simster7) [inline docs/usage-indicator.md]: This appears to be malformed. How is an `nvidia.com/gpu` unit defined? How do we arrive at this conclusion?

File: docs/usage-indicator.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+## Limitations & Assumptions
+
+To calculate the usage we assume that request/limit/default for a resource is a good enough representative of the pods average usage.
```
Comment (simster7) [inline docs/usage-indicator.md]: I would say this is not a good assumption in general, or even as a way to estimate usage

File: util/usage/indicator.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+		summaries[c.Name] = containerSummary{ResourceList: summaries[c.Name].ResourceList, ContainerState: c.State}
+	}
+	i := wfv1.UsageIndicator{}
+	for _, s := range summaries {
+		duration := s.duration(now)
+		for r, q := range s.ResourceList {
+			i = i.Add(wfv1.UsageIndicator{r: wfv1.NewResourceUsageIndicator(time.Duration(q.Value() * duration.Nanoseconds() / resourceDenominator(r).Value()))})
+		}
+	}
+	return i
```
Comment (simster7) [inline util/usage/indicator.go]: I see some issues with this file -- let's discuss in person?

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 8]
```diff
@@ -858,6 +859,46 @@ type RetryStrategy struct {
 	Backoff *Backoff `json:"backoff,omitempty" protobuf:"bytes,3,opt,name=backoff,casttype=Backoff"`
 }
 
+// An indicator (i.e. indicative but not accurate) amount of resource * time in seconds, e.g.
+// CPU 1000m * 1m = 1m
+// memory 1Gi * 2m = 2Gi
+// this is represented as duration in seconds, so can be converted to and from duration (with loss of precision below seconds)
+type ResourceUsageIndicator int64
```
Comment (simster7) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]: Seems like all this does is store seconds? Why is it called a `ResourceUsageIndicator`?

File: util/usage/indicator_test.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	wfv1 "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+)
+
+func TestIndicatorForPod(t *testing.T) {
+	now := time.Now()
+	zero := now.Add(-time.Since(now))
```
Comment (simster7) [inline util/usage/indicator_test.go]: Why isn't `zero` `now` here? I.e. why do we need a zero as opposed to just using now?

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 10]
```diff
@@ -858,6 +859,46 @@ type RetryStrategy struct {
...
+func (i ResourceUsageIndicator) String() string {
+	return i.Duration().String()
+}
+
+// this represents a usage summary by resource (e.g. "memory", "cpu")
+type UsageIndicator map[apiv1.ResourceName]ResourceUsageIndicator
+
+func (i UsageIndicator) Add(o UsageIndicator) UsageIndicator {
+	for r, d := range o {
+		i[r] = i[r] + d
```
Comment (simster7) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]:
```suggestion
		i[r] += d
```
?

File: util/usage/indicator.go [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+func IndicatorForPod(pod *corev1.Pod, now time.Time) wfv1.UsageIndicator {
+	summaries := map[string]containerSummary{}
+	for _, c := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+		summaries[c.Name] = containerSummary{ResourceList: map[corev1.ResourceName]resource.Quantity{
+			// https://medium.com/@betz.mark/understanding-resource-limits-in-kubernetes-cpu-time-9eff74d3161b
+			corev1.ResourceCPU: resource.MustParse("100m"),
+			// https://medium.com/@betz.mark/understanding-resource-limits-in-kubernetes-memory-6b41e9a955f9
+			corev1.ResourceMemory: resource.MustParse("100Mi"),
+		}}
+		for n, q := range c.Resources.Limits {
```
Comment (simster7) [inline util/usage/indicator.go]: Not obviously clear what `n`, and `q` are here. Could we use full variable names?

**Comment by alexec** [comment]
@simster7 I spoke to @jessesuen and he thought better naming would improve things. Rather than act like this is a usage indicator, say what it is - "resources requested * duration". So I've made that change. Thoughts?

**Comment by simster7** [comment]
@alexec That would make more sense to me. I still think this should be opt-in

**Comment by alexec** [comment]
@jessesuen suggested the same. Some kind of feature flag..

**Comment by mac9416** [comment]
I was going to take this PR for a spin. But after building it, I'm getting this from the UI:

```
Request has been terminated Possible causes: the network is offline, Origin is not allowed by Access-Control-Allow-Origin, the page is being unloaded, etc.
```

I'm not getting the error from master. Any ideas?

**Comment by mac9416** [comment]
I was on `cost` instead of `usage`. Disregard.  :-)

**Comment by mac9416** [comment]
@alexec we're hoping to incorporate this into an upcoming release of a downstream project. I can create a branch to work on some of the cosmetic changes I mentioned. Is there anything else you'd like me to work on while I'm in there?

**Comment by alexec** [comment]
@jessesuen thoughts? I think we should aim for v2.7. This has feature flag as requested.

**Comment by mac9416** [comment]
@jessesuen I'm happy to contribute any changes you need. Also happy to rebase to help free up Alex's time.

**Pull request merged**